### PR TITLE
SIMPLYE-261 - Municipality and consortium management

### PR DESCRIPTION
**Pull request by attemoi**
### Goal

- Possibility to customize lanes for each Kimppa (aka consortium)
- All Kimppas share the same collections, i.e.:
    -> License pools shared between all Kimppas
    -> All patrons will have access to the same works regardless of Kimppa
- Easy handling for municipality and Kimppa data and mapping. Mapping done based on municipality code received during sign in.

### Available data

1) Municipality code for patron and admin during authentication
2) Two 3rd party APIs:
    - Kirkanta, <https://api.kirjastot.fi/>
        - Provides list of cities and Kimppas
    - Kansallinen koodistopalvelu, <https://koodistopalvelu.kanta.fi/codeserver/pages/classification-view-page.xhtml?classificationKey=362&versionKey=440>
        - Provides list of municipality codes and names. The authentication system uses these codes to define the users' municipality

### Solution

- One library per Kimppa (+ default library) configured in circulation
  - Each library configured with list of member municipalities
- Patrons and admins automatically assigned to library during sign in
- Unauthenticated users will see the default library lanes
- Collections defined for default library apply to all other libraries
- New monitor script (`ekirjasto_consortium_monitor`) to automatically keep libraries (i.e. Kimppas) synced with Kirkanta data

### Other notes

- No database changes needed
- No changes to authorization logic for list/lane management
- Some municipalities don't belong to a Kimppa
- Users with "turvakielto" have municipality code `000`, i.e. they won't belong to any Kimppa

## Motivation and Context

https://docs.google.com/document/d/1oq9_YBgfY5JMh_zDROZyDHts5FZfH_PQ1FRz1IVxv-k/edit#heading=h.y5cwsxdy2zq5

<https://jira.lingsoft.fi/browse/SIMPLYE-261> - Prepare the data foundation to handle municipality and library patron info 
<https://jira.lingsoft.fi/browse/SIMPLYE-262> - Implement the mapping between municipalities and libraries 
<https://jira.lingsoft.fi/browse/SIMPLYE-263> - Implement custom lane management for local librarians

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

File: api/admin/controller/sign_in.py [hunk 1]
```diff
@@ -94,7 +96,9 @@ def ekirjasto_auth_finish(self):
         if isinstance(user_info, ProblemDetail):
             return user_info
 
-        circulation_role = self._to_circulation_role(user_info.role)
+        circulation_role = self._to_circulation_role(
+            self._db, user_info.role, user_info.municipality
```
Comment (attemoi) [inline api/admin/controller/sign_in.py:100]: This part is still pending some discussion/decisions on how the two municipality fields (`municipality` and `municipalities`) we get during authentication (with user info) will be managed.

Comment (attemoi) [reply api/admin/controller/sign_in.py]: This is now handled. The singular `municipality` field will be used for patrons and `municipalities` field for admins and librarians to determine which libraries they can manage.

**Comment by sylvain-degeilh-lingsoft** [comment]
Some updated information:
- Municipalities without a kimppa should be set into their own 1-municipality kimppa.
- Patrons with turvakielto should not get access to the service at the moment. Loikka is expected to handle that aspect, as well as patrons from municipalities not taking part in the E-kirjasto service.

**Comment by attemoi** [comment]
@sylvain-degeilh-lingsoft, thanks for the update!

> Municipalities without a Kimppa should be set into their own 1-municipality kimppa.

This one is already supported, as we can manage a library for those municipalities manually, without automatic sync. If needed, can be automated further with a separate PR.

**Review by sylvain-degeilh-lingsoft — APPROVED** [review]

**Pull request merged**